lib/grpmat.gi

```diff
@@ -1241,6 +1241,146 @@ InstallGlobalFunction( "BlowUpIsomorphism", function( matgrp, B )
+    D := [];
+    for i in [1..d] do
+        M := IdentityMatrix(GF(q),d);
+        M[i][i] := z;
+        if ConjugateGroup(G,M) = G then
+            M := ConjugatorAutomorphismNC(G,M);
+            Add(D,M);
+        fi;
+    od;
+    D := GroupWithGenerators(D);
+    return D;
+end);
+
+#############################################################################
+##
+#F  FieldAutomorphismGroupOfSL(<G>)
+## 
+##  Returns the group of all field automorphsims of G = SL(d,q).
+##  A field automorphism of SL(d,q), say f : G -> G is given as
+##  f(U) = [sigma(uij)]_{ij} where uij is an entry in U and sigma
+##  is an automorphism over the field that G acts over.
+##
+InstallGlobalFunction("FieldAutomorphismGroupOfSL",function(G)
+    local f,F,applyf,gens,fgens,d,q;
+    d := SLDegree(G);
+    q := Size(SLUnderlyingField(G));
+    f := FrobeniusAutomorphism(GF(q));
+    # f can generate all automorphisms of GF(q)
+    applyf := function(M)
+        local Mnew,i,j;
+        Mnew := List(M,ShallowCopy);
+        for i in [1..d] do
+            for j in [1..d] do
+                Mnew[i][j] := Image(f,M[i][j]);
+            od;
+        od;
+        return Mnew;
+    end;
+    gens := GeneratorsOfGroup(G);
+    fgens := List(gens,applyf);
+    F := GroupHomomorphismByImagesNC(G,G,gens,fgens);
+    F := GroupWithGenerators([F]);
+    return F;
+end);
+
+#############################################################################
+##
+#F  GraphAutomorphismOfSL(<G>)
+##
+##  Returns the graph automorphism g : G -> G of G = SL(d,q),
+##  given by g(U) = (U^T)^-1 for all U in G.
+##
+InstallGlobalFunction("GraphAutomorphismOfSL",function(G)
+    local g,gens,ggens,d,q;
+    d := SLDegree(G);
+    q := Size(SLUnderlyingField(G));
+    gens := GeneratorsOfGroup(G);
+    ggens := List(gens, U -> (TransposedMat(U))^-1);
+    g := GroupHomomorphismByImagesNC(G,G,gens,ggens);
+    return g;
+end);
+
+
+#############################################################################
+##
+#F  SLAutomorphismDecomposition(<G>,<alpha>)
+##
+##  Returns the decomposition of automorphism alpha of G
+##  (a special linear group) as
+##  alpha = i * d * g * f , or
+##  alpha = i * d * e * f where
+##  i is an inner automorphism of G,
+##  d is a diagonal automorphism of G,
+##  f is a field automorphism of G,
+##  g : G \to G is the graph automorphism i.e. g(U) = (U^T)^-1, and
+##  e = g^2 is the identity automorphism.
+##
+InstallGlobalFunction("SLAutomorphismDecomposition",function(G,alpha)
+    local D,F,I,g,gg,dinv,finv,ginv,i,d,f,gfinv,dgfinv;
+    D := DiagonalAutomorphismGroupOfSL(G);
+    F := FieldAutomorphismGroupOfSL(G);
+    I := InnerAutomorphismGroup(G);
+    g := GraphAutomorphismOfSL(G);
+    gg := GroupWithGenerators([g]);
+    for finv in F do
+        for ginv in gg do
+            gfinv := finv * ginv;
+            for dinv in D do
+                dgfinv := gfinv * dinv;
+                i := alpha * dgfinv;
+                if i in I then
```

Comment:
This function is useless for anything but toy examples. (A better method might start by checking how alpha acts on a scalar matrix to determine the field automorphism part. Then divide that out of `alpha`, so that `alpha` fixes the base field. Now by observing how `alpha` maps a set of well-chosen transvections and/or reflections you can figure out a base change matrix `M` that cancels out the inner and diagonal part -- so either the automorphism corresponding to `M` is equal to alpha, or they differ by a graph automorphism. The final step would be to split `M` into an "inner" part (i.e. a matrix in `SL_n`) and a diagonal part: just compute `z:=det(M)` for that. Then you can divide out a diagonal matrix `diag(z,1,...,z)` of `M` to get the desire decomposition muuuch quicker

But it is also unclear who would need this and for what?

Reply:
I did not fully understand your idea.
You say that we can find the images of scalar matrices to find the field automorphism. I understand that the effect of the inner and diagonal parts on a scalar matrix is nothing and it would give us the image of the scalar matrix (let's say $x\cdot I$) under the automorphism $g\circ f$ , or just $f$ if the graph automorphism is not in the decomposition, i.e. it will give $(f(x))^-1 \cdot I$ or $f(x)$ . Apart from figuring out which of these cases we are dealing with, we can do only for $x$ such that $x^d = 1$ (or $Z(q)^0$ if you prefer that), so we can't construct the field automorphism $f$ using `RingHomomorphismByImages` or similar functions since we might not have the image of each generator in the ring. Then how will we divide it out of $\alpha$ ?
I'm sorry if I'm asking a dumb question. I too want to speed up this function.

Reply:
Seems I missed this comment, sorry.

A scalar matrix is in the center of your group, so all inner automorphisms fix it. But  in fact also diagonal automorphism fix scalar matrices. But actually I made a mistake, the graph/diagram automorphism indeed acts by inversion. But that's fine! because inversion is not a field automorphism... 

So what you do is you pick a primitive root $z$ of the finite field and consider the image of the matrix $zI$  under your group automorphism. It must be another scalar matrix, say $f(zI) = z^k I$. Now you can make a case distinction: either $k$ is a power of $p$ (the characteristic of the field), then there is no graph automorphism and the field automorphism is given by $z \mapsto z^k$. Otherwise there is a graph automorphism, and the field automorphism is given by $z \mapsto z^{-k}$ (and $(q-1)-k$ is a power of $p$).